### DISCUSSION TITLE
Make Matplotlib and Plotly optional

### DISCUSSION THREAD

**Discussion by firai** [zauberzeug/nicegui#1010] [Ideas / Feature Requests]
I would like to request making Matplotlib and Plotly optional. If I'm using Plotly, I want to avoid dragging along Matplotlib and all of its dependencies, and vice versa. Sometimes I don't need either.

**Comment by sergeyyurkov1** [comment]
You can do that with the `exclude` parameter in `ui.run`

https://nicegui.io/documentation/run

**Comment by firai** [comment]
I meant as Python installation dependencies. These packages don't seem to be central to NiceGui's functionality.

**Comment by rodja** [comment]
> I meant as Python installation dependencies.

Ah so your major concern is package installation size?

**Comment by firai** [comment]
Right. I could do something hacky and create dummy `setup.py` files for these packages, but I'm not sure whether that might lead to unpredictable behavior.

**Comment by CatamountJack** [comment]
I would second this - they are great for those that use them, but they are pretty heavy dependencies when none of my applications are likely to use either of them.

**Comment by rodja** [comment]
@falkoschindler Will that improve with 1.3 where we load dependencies only when really needed by a ui.element?

**Reply by falkoschindler** [reply]
@rodja Version 1.3 will reduce the amount of JavaScript that is loaded with every page request. But it doesn't reduce NiceGUI's package size and the amount of Python dependencies. The fact that plotly is included in the NiceGUI pip package is probably less of a concern than the matplotlib dependency, which in turn introduces other dependencies.

**Reply by firai** [reply]
Even Plotly is large relative to the rest of the dependencies. As installed on Windows with pip, NiceGui with all other dependencies, i.e., without MPL and Plotly, currently take up 44 MB. MPL (and the pip version of its largest dependency numpy) adds 116 MB, while Plotly adds another 88 MB.

| Configuration                   | Size[^1]     |
| -------------------------------- | -------- |
| Regular installation of NiceGui         | 248.1 MB |
| NiceGui w/ only MPL[^2]             | 160.6 MB |
| NiceGui w/ only Plotly[^2]          | 132.6 MB |
| NiceGui w/o MPL or Plotly[^2]       | 44.2 MB |

[^1]: Differential size from relatively minimal portable WinPython environments. Packages installed with pip. Using `conda`/`miniconda`/`micromamba` environments may yield different results (much larger sizes for MPL if MKL numpy gets installed from `conda` channel).
[^2]: Achieved by installing NiceGui with `--no-deps` and then manually installing dependencies.

**Comment by rodja** [comment]
I've just created pull request https://github.com/zauberzeug/nicegui/pull/1143 which makes these large packages optional.

**Comment by CatamountJack** [comment]
Firstly, this change is great - just tried it with 1.3.2 and it saved 38+ MB off my packaged app.  Awesome!! 👍 👍 Thanks @rodja and @falkoschindler and thanks @firai for starting the discussion!!

With that said... probably my ignorance, but by making these "optional", how would a user know that they need to install these additional dependencies manually?  I didn't notice anything in the documentation.  I started from scratch in my virtualenv with 1.3.2 and when I ran my app with native mode turned on I got a cryptic error `AttributeError: 'WindowProxy' object has no attribute 'signal_server_shutdown'`.  I realized immediately it was probably because `pywebview` wasn't installed, but I'm not sure a new user would.

**Reply by falkoschindler** [reply]
For missing matplotlib and plotly NiceGUI throws an import error. But you're right, something like this is missing for pywebview. And we should add notes about extra dependencies to the documentation.

**Comment by firai** [comment]
Thanks @rodja and @falkoschindler! I'll close the discussion as the request has been fulfilled. Issues regarding import errors, fine-tuning of import logic or documentation can be dealt with by filing specific issues.
